### PR TITLE
3036 tests add fillet

### PR DESCRIPTION
**Pull request by max-mrgrsk**
## Implemented tests for getPathToExtrudeForSegmentSelection:

This PR adds tests to verify that the function correctly identifies extrude paths based on selected segments across different sketches. The tests cover multiple scenarios, including cases with multiple extrusions and situations where no valid extrusion is present. This helps ensure the function behaves correctly and handles complex cases.

Additionally, for the sake of testing, the `addFillet` file was heavily refactored. All fillet-related logic was moved from the `modellingMachine` file to `addFillet`, making the code more modular and easier to test.

Related Issue: #3036 (from #2606: list of stuff for fillets)

### CODE REVIEW

**Comment by max-mrgrsk** [comment]
@Irev-Dev could you please have a look on errors in the `addFillet.test.ts` and test cases overall ?

**Comment by max-mrgrsk** [comment]
### Update: 
Adjusted the initialization logic to ensure that engineCommandManager connects properly. Same as in the `artifactGraph.test.ts`

File: src/lang/modifyAst/addFillet.ts [hunk 1]
```diff
@@ -64,120 +190,101 @@ export function addFillet(
...
-
   if (extrudeInit.type === 'CallExpression') {
     // 1. no fillet case
     extrudeDeclarator.init = createPipeExpression([extrudeInit, filletCall])
     return {
-      modifiedAst: _node,
+      modifiedAst: node,
       pathToFilletNode: getPathToNodeOfFilletLiteral(
         pathToExtrudeNode,
         extrudeDeclarator,
```
Comment (Irev-Dev) [inline src/lang/modifyAst/addFillet.ts:290]: This function is getting a bit confusing because you need to know how mutating structures effect each other.

so here we mutate one object `extrudeDeclarator`
`extrudeDeclarator.init = createPipeExpression([extrudeInit, filletCall])`
And then immediately return another object `node`
`return { modifiedAst: node, ...`
And the reason this works is because `extrudeDeclarator` is actually a child/part of `node` so by mutating the former we are also mutating the later. The problem with this is that's completely opaque to someone reading and trying to understand this function for the first time, and in fact the function won't work as intended if for some reason someone else used this function and passed in  a `extrudeDeclarator` that was not actually part of `node` (this isn't an exported function so that's unlikely to happen, but the readability still stands)


 I think breaking this all into so many functions here is a bit of a mixed blessing, while lots of functions can help readability from the sense of getting the narrative of what's happening, it does abstract you away from the nuts and bolt and when the nuts and bolt is mutating two variables that are actually implicitly connected it's a problem.
 
In cases of ast mods, queries can be broken out in helper functions, but mutations probably need to stay in the one function so that the whole story is contained in that function.

My apologies because I know I pointing this out now, when some of these abstractions were done in previous PRs.

Does what I'm saying make sense?

Comment (Irev-Dev) [reply src/lang/modifyAst/addFillet.ts:290]: Been thinking about this comment a bit more, I don't like that its so high level without concrete examples of what I mean particularly:

> In cases of ast mods, queries can be broken out in helper functions, but mutations probably need to stay in the one function so that the whole story is contained in that function.

I think this is an hypothetical function that I think mixes well both adding abstractions while none of the mutations leave the main function:

```ts
function hyotheticalDoStuff(
  ast: Program,
  pathToNode1: PathToNode,
  pathToNode2: PathToNode,
): {modifiedAst: Program} | Error {
  const astClone = structuredClone(ast)

  const node1 = getNodeFromPath<CallExpression>(astClone, pathToNode1, 'CallExpression')
  const node2 = getNodeFromPath<VariableDeclaration>(astClone, pathToNode2, 'VariableDeclaration')

  if(err(node1)) return node1
  if(err(node2)) return node2

  // any bussiness logic checks for these nodes
  const isNode1Good = checkNode1Specifics(node1)
  const isNode1Good = checkNode2Specifics(node1)

  if (err(isNode1Good)) return isNode1Good
  if (!isNode1Good) return new Error('Node 1 is not good')

  // lets say we need to update the callexpressions arguments and the variable declarations init
  // we have a helper function that creates the updated node, but does not mutate anything
  const updatedChildeNodePart1 = updateChildNode1(node1.node)
  const updatedChildeNodePart2 = updateChildNode2(node1.node)

  // we wait till we're back to the main thread to apply the changes
  node1.node.arguments[0] = updatedChildeNodePart1
  node2.node.declarations[0].init = updatedChildeNodePart2

  // this way when reading this function we can read the "story" that node1 and node2 are children that come from the astClone object
  // and that by mutating them we're also mutating the astClone object and that's why we're returning astClone.
  // none of the mutations are tucked away in the helper functions, if they were then it would be using an input param as the output, which we want to avoid.

  return {modifiedAst: astClone}
}
```

I really hope I'm making sense with all this.

File: src/lang/modifyAst/addFillet.ts [hunk 2]
```diff
@@ -231,6 +322,89 @@ export function addFillet(
...
+    if (prop.key.name === 'tags' && prop.value.type === 'ArrayExpression') {
+      return prop.value.elements.some(
+        (element) => element.type === 'Identifier' && element.name === tag
+      )
+    }
+    return false
+  })
+}
+
+function getPathToRadiusLiteral(node: ObjectExpression, path: any): any {
```
Comment (Irev-Dev) [inline src/lang/modifyAst/addFillet.ts:375]: Is this right?
```suggestion
function getPathToRadiusLiteral(node: ObjectExpression, path: any): PathToNode {
```

File: src/lang/modifyAst/addFillet.ts [hunk 3]
```diff
@@ -231,6 +322,89 @@ export function addFillet(
   return new Error('Unsupported extrude type.')
 }
 
+function getPathToNodeOfFilletLiteral(
+  pathToExtrudeNode: PathToNode,
+  extrudeDeclarator: VariableDeclarator,
+  tag: string
+): PathToNode {
+  let pathToFilletObj: any
```
Comment (Irev-Dev) [inline src/lang/modifyAst/addFillet.ts:330]:
```suggestion
  let pathToFilletObj: PathToNode
```
Or maybe needs to be
```suggestion
  let pathToFilletObj: PathNoNode = []
```

File: src/lang/modifyAst/addFillet.ts [hunk 4]
```diff
@@ -231,6 +322,89 @@ export function addFillet(
...
+        ['properties', 'ObjectExpression'],
+        [index, 'index'],
+        ['value', 'Property']
+      )
+    }
+  })
+  return pathToFilletObj
+}
+
+function getFilletTag(existingFilletCall: CallExpression): string | null {
```
Comment (Irev-Dev) [inline src/lang/modifyAst/addFillet.ts:389]: `export function getTagFromCallExpression(` in `src/lang/std/sketch.ts` has a very similar function signature, just wanted to double check if these are doing different things?

**Review by Irev-Dev — COMMENTED** [review]
Sorry left a few comments. Let me know what you think.

**Comment by max-mrgrsk** [comment]
> Sorry left a few comments. Let me know what you think.

Comments are pure gold, thank you! I will start working on them tonight and hopefully tackle over the weekend :)

File: src/lang/modifyAst/addFillet.ts [hunk 5]
```diff
@@ -28,26 +29,151 @@ import {
...
+ * Apply Fillet To Selection
+ */
+
+export function applyFilletToSelection(
+  selection: Selections,
+  radius: KclCommandValue
+): void | Error {
+  // 1. get AST
+  let ast = kclManager.ast
+  const astResult = getAst(ast, radius)
```
[the remark this answers is not in the document]
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts:52]: good point, thanks. What do you think about mutateAstForRadiusInsertion?

[the remark this answers is not in the document]
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts:52]: though insertRadiusIntoAst fits better.. what do you think?

File: src/lang/modifyAst/addFillet.ts [hunk 6]
```diff
@@ -28,26 +29,151 @@ import {
...
+  try {
+    // Validate and update AST
+    if (
+      'variableName' in radius &&
+      radius.variableName &&
+      radius.insertIndex !== undefined
+    ) {
+      const newBody = [...ast.body]
+      newBody.splice(radius.insertIndex, 0, radius.variableDeclarationAst)
+      ast.body = newBody
```
[the remark this answers is not in the document]
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts:97]: thanks, fixed! 4a44ee4fa44a5564bbf5549b02ea87aff3371aa5

File: src/lang/modifyAst/addFillet.ts [hunk 7]
```diff
@@ -28,26 +29,151 @@ import {
...
+
+  const sketchGroup = programMemory.get(sketchVar)
+  if (sketchGroup?.type !== 'SketchGroup')
+    return new Error('Invalid sketch group type')
+
+  const extrusion = getExtrusionFromSuspectedPath(sketchGroup.id, artifactGraph)
+
+  const pathToExtrudeNode = err(extrusion)
+    ? []
+    : getNodePathFromSourceRange(ast, extrusion.codeRef.range)
```
[the remark this answers is not in the document]
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts]: fixed, thanks

File: src/lang/modifyAst/addFillet.ts [hunk 8]
```diff
@@ -28,26 +29,151 @@ import {
...
+/**
+ * Add Fillet
+ */
 
 export function addFillet(
   node: Program,
   pathToSegmentNode: PathToNode,
   pathToExtrudeNode: PathToNode,
+  // radius = createLiteral(5) as Expr // TODO what is EXP?
   radius = createLiteral(5) as Expr
```
[the remark this answers is not in the document]
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts]: thanks, good point.. I've used `radius: Expr = createLiteral(5)` . Just deleting `as Expr` did not work because ts is not sure whether radius is a `Literal` in this line: `'variableName' in radius ? radius.variableIdentifierAst : radius.valueAst` . efbc67878bc4ca485d087aa9f6e2cf4f803cdc01

File: src/lang/modifyAst/addFillet.ts [hunk 9]
```diff
@@ -64,120 +190,101 @@ export function addFillet(
...
+    pathToExtrudeNode,
+    tag
+  )
+}
+
+function locateExtrudeCall(
+  node: Program,
+  pathToExtrudeNode: PathToNode
+): { extrudeDeclarator: VariableDeclarator; extrudeInit: Expr } | Error {
+  const extrudeChunk = getNodeFromPath(
```
[the remark this answers is not in the document]
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts]: thanks, fixed: da81df48d892d1489e833ef5215469702d3631b0

File: src/lang/modifyAst/addFillet.ts [hunk 10]
```diff
@@ -64,120 +190,101 @@ export function addFillet(
...
+    node,
     pathToExtrudeNode,
     'VariableDeclaration'
   )
   if (err(extrudeChunk)) return extrudeChunk
-  const { node: extrudeVarDecl } = extrudeChunk
 
+  const { node: extrudeVarDecl } = extrudeChunk as {
+    node: VariableDeclaration
+  }
```
[the remark this answers is not in the document]
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts:252]: fixed

File: src/lang/modifyAst/addFillet.ts [hunk 11]
```diff
@@ -64,120 +190,101 @@ export function addFillet(
...
-      extrudeInit.type !== 'PipeExpression')
+    extrudeInit.type !== 'CallExpression' &&
+    extrudeInit.type !== 'PipeExpression'
   ) {
-    return new Error('Extrude PipeExpression / CallExpression not found.')
+    return new Error('Extrude must be a PipeExpression or CallExpression')
   }
 
-  // determine if extrude is in a PipeExpression or CallExpression
+  const extrudeExpressionType = extrudeInit.type
```
[the remark this answers is not in the document]
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts:270]: right, sorry

File: src/lang/modifyAst/addFillet.ts (hunk 1 above)
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts:293]: this is the coolest comment ever, I'll grab a ~~beer~~ cup of tea and read it a couple times more.

Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts:293]: My goal was to break down the complex logic into simple pieces—`addFillet` tags the segment and adds the fillet, and the helpers handle the details. 

Complex shit is just piles of simple tasks, so I organized them that way.

But I see your point about keeping AST mutations together for clarity. Yeah, I totally missed the moment that mutations are not obvious.

File: src/lang/modifyAst/addFillet.ts (hunk 4 above)
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts:401]: it has different purpose and structure (for segments) and has to be heavily modified in order to work with fillets. 

```
export function getTagFromCallExpression(
  callExp: CallExpression
): string | Error {
  if (callExp.callee.name === 'close') return getTag(1)(callExp)
  if (callExp.callee.name in sketchLineHelperMap) {
    const { getTag } = sketchLineHelperMap[callExp.callee.name]
    return getTag(callExp)
  }
  return new Error(`"${callExp.callee.name}" is not a sketch line helper`)
}
```

so we would have add another if statement like: `if (callExp.callee.name === 'fillet')` and then copy paste the logic of current `getFilletTag` in it.

Or you mean using the same logic as over there and do something similar?
would be something like this:
```
function getTagFromCallExpression2(callExp: CallExpression): string | Error {
  return getTag()(callExp)
}

function getTag(index = 0) {
  return (callExp: CallExpression) => {
    if (callExp.type !== 'CallExpression') {
      return new Error('Not a CallExpression')
    }

    // Access the correct argument based on the provided index
    const arg = callExp.arguments?.[index]
    if (!arg || arg.type !== 'ObjectExpression') {
      return new Error('Expected an ObjectExpression as the argument')
    }

    const properties = (arg as ObjectExpression).properties
    const tagsProperty = properties.find((prop) => prop.key.name === 'tags')
    
    if (!tagsProperty || tagsProperty.value.type !== 'ArrayExpression') {
      return new Error('No valid tags property found')
    }

    const elements = (tagsProperty.value as ArrayExpression).elements
    if (elements.length === 0 || elements[0].type !== 'Identifier') {
      return new Error('No valid tag identifier found')
    }

    return elements[0].name // Return the name of the first tag (identifier)
  }
}
```

File: src/lang/modifyAst/addFillet.ts (hunk 1 above)
Comment (Irev-Dev) [reply src/lang/modifyAst/addFillet.ts]: Thank you, and thanks for linking to the commit.

I'm not sure what you mean by `createPipeExpression` because that function does not mutate anything, it's creating an entirely new data structure.

I should clarify, I'm using mutate here to mean changing underlying memory.

like this would be a function that mutates
```js
const doesMutate = (val) => {
  val.newProperty = 'hi'
  return val
}
```
Where is this does the same thing, but because it does a shallow close, not not effect the underlying memory, because it creates a new object with the `...` spread operator
```js
const doesNotMutate = (val) => {
  const newVal = {...val, newProperty: 'hi}
  return newVal
}
```
I might as well be thorough at risk of making you read stuff you're already aware of.

This relates to much of the idea of "pure function", as a pure function is not supported to effect anything outside of itself (that is no "side effects") and mutating is a side effect because it can effect other parts of the code and lead to tricky bugs.

The thing I was most concerned about that I called out last time was mutating extrudeDeclarator in order to also mutate the ast because `extrudeDeclarator` was a child of the ast, the connection between the two was implicit and very hard to follow without going outside the function and looking at how it was being called. Here's another example of functions params being mutated causing issues, 

image

I've intentionally cut some of setup code from the start of this snippet for the purposes of putting yourself in the shoes of someone trying to understand what the code is doing, I can imagine thinking to myself

> "why did `itGets` log out two different things, when I called it with the exact same pararm? sure I know that `itMutates` mutates, but they use different variables, `itGets` using `yo` and it mutates uses `myAbc`!?"

The full snipped explains it because `yo` and `myAbc` in infact connected, so by muting `myAbc` we're also mutating `yo`

image

In this toy example seeing the connection is very clear because it's just above, but it's get tricky of those two variables get passed along through multiple function, you have to go upstream to realise there's two separate vars are linked.

I think maybe I should clarify what a mutation looks like javascript
for objects if you see `myObj.propertyName =` or `myObj[stringVariable] =` than `myObj` is being mutated
For arrays it's probably more like `.push(), .pop(), .shift(), .unshift()`

Maybe that relates to your comment about `createPipeExpression`, because there's the line `extrudeDeclarator.init = createPipeExpression([...`?

So there's a mutation here but it's not `createPipeExpression`, it's the first half `extrudeDeclarator.init =` where the mutation is happening, but since your change in https://github.com/KittyCAD/modeling-app/commit/045a05ae1086ff5b215240ea3c6eac44327cfafd that mutation is now happening on a variable that's declared in the same function so it's much easier to follow, where as before it was being passed in as an input param.

**Review by Irev-Dev — APPROVED** [review]

**Comment by Irev-Dev** [comment]
Merge it!

Before you do you will need to make the linter happy

image


This is probably not your fault, changing CI to not allow any warning has been a recent change, but since there are only two it should be a quick fix.

File: src/lang/modifyAst/addFillet.ts (hunk 1 above)
Comment (max-mrgrsk) [reply src/lang/modifyAst/addFillet.ts]: I think I need to develop a habit of printing out your comments and hanging them on the wall. in the frame :D thank you!
